connector/src/main/scala/com/google/cloud/spark/bigquery/BigQueryStreamWriter.scala

```diff
@@ -39,17 +36,13 @@ private[bigquery] object BigQueryStreamWriter extends Logging {
                  outputMode: OutputMode,
                  opts: SparkBigQueryConfig,
                  bigQueryClient: BigQueryClient): Unit = {
-    val schema: StructType = data.schema
-    val expressionEncoder: ExpressionEncoder[Row] = RowEncoder(schema).resolveAndBind()
-    val rowRdd: RDD[Row] = data.queryExecution.toRdd.mapPartitions(iter =>
-      iter.map(internalRow =>
-        expressionEncoder.fromRow(internalRow)
-      )
-    )
-    // Create fixed dataframe
-    val dataFrame: DataFrame = sqlContext.createDataFrame(rowRdd, schema)
-    val table = Option(bigQueryClient.getTable(opts.getTableId))
 
+    // using createDataFrame(List[Row], schema) instead of
+    // createDataFrame(RDD[Row], schema) as for creating RDD[Row] we need to use
+    // ExpressionEncoder, which has breaking changes from spark 2.x to 3.x
+    val dataFrame = sqlContext.createDataFrame(data.collectAsList(), data.schema)
```

Comment:
`collectAsList()` collects all the data to the driver, and as such it cannot be used for large DataFrames. We need another approach that work on the data in the partitions

Reply:
Based on https://github.com/apache/spark/commit/fab4ca5156d5e1cc0e976c7c27b28a12fa61eb6d I'd try the `ExpressionEncoder.createDeserializer()`. Notice you can have different approach per spark version.